src/curvefit/uncertainty/draws.py

```diff
@@ -0,0 +1,140 @@
+        self.exp_smoothing = exp_smoothing
+        self.max_last = max_last
+        self._draws = None
+
+        assert type(self.num_draws) == int
+        assert self.num_draws > 0
+
+        assert type(self.prediction_times) == np.ndarray
+
+        if self.exp_smoothing is not None:
+            assert type(self.exp_smoothing) == float
+            if self.max_last is None:
+                raise RuntimeError("Need to pass in how many of the last models to use.")
+            else:
+                assert type(self.max_last) == int
+                assert self.max_last > 0
+        else:
+            if self.max_last is not None:
+                raise RuntimeError("Need to pass in exponential smoothing parameter.")
+
+    # TODO: Implement exponential smoothing (do we need it?)
+    def create_draws(self,
+                     data: Data,
+                     model_prototype,
+                     solver_prototype,
+                     residual_model: _ResidualModel,
+                     evaluation_space,
+                     theta: float):
+
+        self._draws = {group: [] for group in data.groups}
+        for group in data.groups:
+
+            # Initializing model and doing the "best fit"
+            model = model_prototype.clone()
+            solver = solver_prototype.clone()
+            solver.set_model_instance(model)
+            current_group_data = data._get_df(group=group, copy=False)
+            solver.fit(current_group_data)
+
+            # getting predictions for draws times
+            predictions = solver.predict(
+                t=self.prediction_times,
+                predict_fun=evaluation_space
+            )
+
+            max_t = int(np.round(current_group_data[data.col_t].max()))
+            num_obs = int(sum(~current_group_data[data.col_obs].isnull()))
+
+            add_noise = self.prediction_times > max_t
+            no_noise = self.prediction_times <= max_t
+
+            forecast_out_times = self.prediction_times[add_noise] - max_t
+            error = residual_model.simulate_residuals(
+                num_simulations=self.num_draws,
+                covariate_specs={'num_data': num_obs, 'far_out': forecast_out_times}
+            )
+            no_error = np.zeros(shape=(self.num_draws, sum(no_noise)))
+            all_error = np.hstack([no_error, error])
+
+            noisy_forecast = predictions - (predictions ** theta) * all_error
+
+            noisy_forecast = data_translator(
+                data=noisy_forecast, input_space=evaluation_space, output_space=evaluation_space
+            )
+            self._draws[group].append(noisy_forecast)
+
+        return self
+
+    def get_draws(self, group=None):
+        if self._draws is None:
+            raise RuntimeError("Draws are not created yet: call create_draws(..) first.")
+
+        if group is not None:
+            return self._draws[group]
+        else:
+            return self._draws
```

Comment:
One last function that I think we should add is a way to get the mean and quantiles of the draws. Could either build it into the `get_draws` function or maybe make a `get_summaries` function instead? They wouldn't be used in production but for us to make diagnostic plots more easily.

Reply:
Sure! See this commit:
https://github.com/ihmeuw-msca/CurveFit/pull/85/commits/4bb91eba149fb21031a18a2bf598845a07ab15c4

Reply:
Great!